csharp/src/IceLocatorDiscovery/PluginI.cs

```diff
@@ -256,350 +173,224 @@ public void
+                if (t == completionSource.Task)
                 {
-                    Debug.Assert(_pendingRequests.Count == 0);
-                    return; // Request failed
+                    return await completionSource.Task.ConfigureAwait(false);
                 }
 
-                if (_pendingRetryCount > 0)
+                lock (_mutex)
                 {
-                    --_pendingRetryCount;
-                    try
+                    if (_completionSource != null)
                     {
-                        if (_traceLevel > 1)
+                        // Locator lookup timeout and no more retries
+                        if (_traceLevel > 0)
                         {
-                            var s = new StringBuilder("retrying locator lookup:\nlookup = ");
+                            var s = new StringBuilder("locator lookup timed out:\nlookup = ");
                             s.Append(_lookup);
-                            s.Append("\nretry count = ").Append(_retryCount);
-                            if (_instanceName.Length == 0)
+                            if (_instanceName.Length > 0)
                             {
                                 s.Append("\ninstance name = ").Append(_instanceName);
                             }
                             _lookup.Communicator.Logger.Trace("Lookup", s.ToString());
                         }
 
-                        foreach (KeyValuePair<ILookupPrx, ILookupReplyPrx?> l in _lookups)
-                        {
-                            _ = FindLocator(l.Key, l.Value!); // Sent multicast request
-                        }
-                        _timer.Schedule(this, _timeout);
-                        return;
-                    }
-                    catch (Exception)
-                    {
-                    }
-                    _pendingRetryCount = 0;
-                }
-
-                Debug.Assert(_pendingRetryCount == 0);
-                _pending = false;
-
-                if (_traceLevel > 0)
-                {
-                    var s = new StringBuilder("locator lookup timed out:\nlookup = ");
-                    s.Append(_lookup);
-                    if (_instanceName.Length == 0)
-                    {
-                        s.Append("\ninstance name = ").Append(_instanceName);
-                    }
-                    _lookup.Communicator.Logger.Trace("Lookup", s.ToString());
-                }
-
-                if (_pendingRequests.Count == 0)
-                {
-                    Monitor.Pulse(this);
+                        _nextRetry = Time.CurrentMonotonicTimeMillis() + _retryDelay;
+                        _completionSource.SetResult(_voidLocator);
+                        _completionSource = null;
+                        _findLocatorTask = null;
+                    } // Else the completion source was completed by a concurrent reply
+                    Debug.Assert(completionSource.Task.IsCompleted);
```

Comment:
This assert looks wrong.

Reply:
mm if the code riches this point the task was completed in the above block, or by a concurrent reply that happen between `WhenAny` return and the lock was acquired

Reply:
Fixed this whole block was supposed to be outside the loop